docs/src/threadpool.rst

```diff
@@ -32,35 +55,156 @@ Data types
 
 .. c:type:: void (*uv_work_cb)(uv_work_t* req)
 
-    Callback passed to :c:func:`uv_queue_work` which will be run on the thread
-    pool.
+    Callback passed to :c:func:`uv_queue_work` which will be run on
+    the executor.
 
 .. c:type:: void (*uv_after_work_cb)(uv_work_t* req, int status)
 
     Callback passed to :c:func:`uv_queue_work` which will be called on the loop
-    thread after the work on the threadpool has been completed. If the work
-    was cancelled using :c:func:`uv_cancel` `status` will be ``UV_ECANCELED``.
+    thread after the work on the executor has been completed or cancelled.
+    If the work was cancelled using :c:func:`uv_cancel`,
+    then `status` will be ``UV_ECANCELED``.
+
+.. c:type:: uv_executor_t
+
+    Executor type. Use when overriding the default threadpool.
+
+.. c:type:: void (*uv_executor_init_func)(uv_executor_t* executor)
+
+    Called once, the first time :c:func:`uv_executor_queue_work` is called.
+    Do any initialization you want.
+
+.. c:type:: void (*uv_executor_destroy_func)(uv_executor_t* executor)
+
+    Called at some point ???
+
+.. c:type:: void (*uv_executor_submit_func)(uv_executor_t* executor, uv_work_t* req, uv_work_options_t* opts)
 
+    Called when the executor should handle this request.
+
+.. c:type:: int (*uv_executor_cancel_func)(uv_executor_t* executor, uv_work_t* req)
+
+    Called when someone wants to cancel a previously-submit'ed request.
+    Return ``UV_EBUSY`` if you cannot cancel it.
+
+.. seealso:: :c:func:`uv_cancel_t`.
+
+.. c:type:: int (*uv_executor_done_cb)(uv_work_t* req)
+
+    libuv sets this during a successful call to :c:func:`uv_replace_executor`.
+    An executor should invoke this CB once finished with a request.
+    This CB is thread safe.
+
+.. c:type:: int uv_replace_executor(uv_executor_t* executor)
+
+    Replace the default libuv executor with this user-defined one.
+    Must be called before any work is submitted to the default libuv executor.
+    Returns 0 on successs.
+
+.. c:type:: uv_work_options_t
+
+    Options for guiding the executor in its policy decisions.
 
 Public members
 ^^^^^^^^^^^^^^
 
+.. c:member:: void* uv_work_t.reserved[0]
+
+    Space for arbitrary data. libuv does not use this field.
+    This is intended for use by an executor implementation.
+
 .. c:member:: uv_loop_t* uv_work_t.loop
 
     Loop that started this request and where completion will be reported.
     Readonly.
 
+.. c:member:: uv_work_cb uv_work_t.work_cb
+
+    Executor should invoke this, not on the event loop.
+
 .. seealso:: The :c:type:`uv_req_t` members also apply.
 
+.. c:member:: uv_executor_init_func uv_executor_t.init
+
+    Invoked once.
+    Can be NULL.
+
+.. c:member:: uv_executor_destroy_func uv_executor_t.destroy
+
+    Can be NULL.
+
+.. c:member:: uv_executor_submit_func uv_executor_t.submit
+
+    Must be non-NULL.
+
+.. c:member:: uv_executor_cancel_func uv_executor_t.cancel
+
+    Can be NULL.
```

Comment:
Would it make sense to use a different return value for `uv_cancel` (e.g. `ENOSYS`) if the executor doesn’t support cancelling?

Reply:
Good idea.

Reply:
Fixed.